File: src/main/java/toast/client/commands/CommandHandler.java
package toast.client.commands;

import net.minecraft.client.MinecraftClient;
import toast.client.ToastClient;
import toast.client.commands.cmds.*;
import toast.client.utils.Logger;
import toast.client.utils.RandomMOTD;

import java.util.Objects;
import java.util.concurrent.CopyOnWriteArrayList;

public class CommandHandler {
    public static CopyOnWriteArrayList<Command> commands = new CopyOnWriteArrayList<>();

    public void executeCmd(String name, String[] args) {
        boolean notfound = true;
        for (Command command : commands) {
            for (String alias : command.aliases) {
                if (alias.equalsIgnoreCase(name)) {
                    try {
                        if (isDevCancel(command)) {
                            notfound = true;
                            continue;
                        }
                        notfound = false;
                        command.run(args);
                    } catch (Exception err) {
                        err.printStackTrace();
                        Logger.message("Sorry but something went wrong", Logger.ERR, true);
                    }
                }
            }
        }
        if (notfound) {
            Logger.message("Cannot find command " + ToastClient.cmdPrefix + name, Logger.ERR, true);
        }
    }

    public boolean isDevCancel(Command c) {
        return c.isDev() && !ToastClient.devs.contains(Objects.requireNonNull(MinecraftClient.getInstance().player).getDisplayName().asFormattedString());
    }

    public CopyOnWriteArrayList<Command> getCommands() {
        return commands;
    }

    public Command getCommand(String cmd) {
        for (Command command : commands) {
            for (String alias : command.aliases) {
                if (alias.equalsIgnoreCase(cmd)) {
                    return command;
                }
            }
        }
        return null;
    }

    public void initCommands() {
        commands.clear();
        // alphabetical order please
        commands.add(new Bind());
        commands.add(new ClearChat());
        commands.add(new FOV());
        commands.add(new GuiReset());
        commands.add(new Help());
        commands.add(new ListModules());
        commands.add(new Panic());
        commands.add(new Prefix());
        commands.add(new Reload());
        commands.add(new MOTD());
        commands.add(new Save());
        commands.add(new Suffix());
        commands.add(new Set());
        commands.add(new Test());
        commands.add(new Toggle());
    }
}

File: src/main/java/toast/client/commands/CommandHandler.kt
package toast.client.commands

import net.minecraft.client.MinecraftClient
import org.reflections.Reflections
import toast.client.ToastClient
import toast.client.utils.MessageUtil
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Manages the loading and execution of commands
 */
class CommandHandler {
    /**
     * Array containing the instances of all the commands
     */
    var commands: CopyOnWriteArrayList<Command> = CopyOnWriteArrayList<Command>()

    /**
     * Executes the command corresponding to name with args as the argument array
     */
    fun executeCmd(name: String, args: Array<String>) {
        var notfound = true
        val commandIter = commands.iterator()
        while (commandIter.hasNext()) {
            val nextCommand = commandIter.next()
            val aliasIter = nextCommand.aliases.iterator()
            while (aliasIter.hasNext()) {
                val nextAlias = aliasIter.next()
                if (nextAlias.equals(name, ignoreCase = true)) {
                    try {
                        if (isDevCancel(nextCommand)) {
                            notfound = true
                            continue
                        }
                        notfound = false
                        nextCommand.run(args)
                    } catch (err: Exception) {
                        err.printStackTrace()
                        MessageUtil.sendMessage("Sorry but something went wrong", MessageUtil.Color.RED)
                    }
                }
            }
        }
        if (notfound) {
            MessageUtil.sendMessage("Cannot find command ${MessageUtil.CHAT_PREFIX} $name", MessageUtil.Color.RED)
        }
    }

    /**
     * Checks if the command is developer only and if it is checks that the current player is a developer
     */
    fun isDevCancel(c: Command): Boolean {
        return c.isDev && !ToastClient.devs.contains(Objects.requireNonNull(MinecraftClient.getInstance().player)!!.displayName.asFormattedString())
    }

    /**
     * Gets a command from it's name
     */
    fun getCommand(cmd: String?): Command? {
        val commandIter = commands.iterator()
        while (commandIter.hasNext()) {
            val next = commandIter.next()
            for (alias in next.aliases) {
                if (alias.equals(cmd, ignoreCase = true)) {
                    return next
                }
            }
        }
        return null
    }

    /**
     * Loads and initializes all of the commands
     */
    fun initCommands() {
        commands.clear()
        val reflections = Reflections("toast.client.commands")
        val commandClasses =
                reflections.getSubTypesOf(
                        Command::class.java
                )
        val classIter = commandClasses.iterator()
        while (classIter.hasNext()) {
            val command = classIter.next().getConstructor().newInstance()
            commands.add(command)
        }
    }
}